File: bruh.py
from math import sqrt

def inner_product(u, v):
    counter = 0
    for i in range(len(u)):
        counter += u[i]*v[i]
    return counter

def scalar_multiply(u, scalar):
    result = []
    for i in u:
        result.append(i*scalar)
    return result

def vector_subtraction(u, v):
    result_2 = []
    for i in range(len(u)): 
        result_2.append(u[i]-v[i])
    return result_2


def normalize(vector):
    norm = (inner_product(vector, vector))**0.5
    return scalar_multiply(vector, 1/norm)

def gram_schmidt(A):
    orthogonal_basis = [A[0]]
    for i in range(1,len(A)):
        counter = 0
        hats = []
        while counter < i:
            hat_top = inner_product(A[i], orthogonal_basis[counter])
            hat_bottom = inner_product(orthogonal_basis[counter], orthogonal_basis[counter])
            hat_final = scalar_multiply(orthogonal_basis[counter], hat_top/hat_bottom)
            hats.append(hat_final)
            counter += 1
        result = A[i]
        for hat in hats:
            result = vector_subtraction(result, hat)
        orthogonal_basis.append(result)
    
    return orthogonal_basis

def vector_add(u, v):
    result = []
    for i in range(len(u)):
        result.append(u[i]+v[i])
    return result

def LLL(Basis, S=3/4):
    n = len(Basis)
    orthobasis = gram_schmidt(Basis)


    mu = [[0 for _ in range(n)] for _ in range(n)]

    for i in range(n):
        mu[i][i] = 1
        for j in range(i):
            mu[i][j] = inner_product(Basis[i], orthobasis[j])/inner_product(orthobasis[j], orthobasis[j])

    k = 1
    while k < n:
        for j in range(k-1, -1, -1):
            if abs(mu[k][j]) > 1/2:
                tempp = scalar_multiply(Basis[j], round(mu[k][j]))
                Basis[k] = vector_subtraction(Basis[k], tempp)
                orthobasis = gram_schmidt(Basis)
                for i in range(n):
                    for p in range(len(mu[i])-1):
                        mu[i][p] = inner_product(Basis[i], orthobasis[p])/inner_product(orthobasis[p], orthobasis[p])
        if inner_product(orthobasis[k], orthobasis[k]) > (S-((mu[k][k-1]))**2)* inner_product(orthobasis[k-1], orthobasis[k-1]):
            k += 1
        else:
            temp = Basis[k]
            Basis[k] = Basis[k-1]
            Basis[k-1] = temp
            orthobasis = gram_schmidt(Basis)
            for i in range(n):
                    for z in range(len(mu[i])-1):
                        mu[i][z] = inner_product(Basis[i], orthobasis[z])/inner_product(orthobasis[z], orthobasis[z])
            k = max(k-1, 1)

    return Basis


def schnorr_euchner(A, R):
    B = gram_schmidt(A)
    n = len(A)
    B_inner = []
    for i in range(len(B)):
        B_inner.append(inner_product(B[i], B[i]))

    mu = [[0 for i in range(n)] for i in range(n)]
    for i in range(len(mu)):
        for j in range(i):
            mu[i][j] = inner_product(A[i], B[j])/inner_product(B[j], B[j])

    p = [0] * (n+1)
    v = [0] * n
    v[0] = 1
    c = [0] * n
    w = [0] * n
    k = 0
    last_nonzero = 0
    R_squared = R**2
    s = [A[0]]

    while True:
        p[k] = p[k+1] + (v[k] + c[k])**2 * B_inner[k]
        
        if p[k] < R_squared: # Bu eşittirin amına koydum
            if k == 0:
                R_squared = p[k]
                s = [0 for _ in range(len(A[0]))]
                for i in range(n):
                    s = vector_add(scalar_multiply(A[i], v[i]), s)
                                
            else:
                k -= 1
                c[k] = 0
                for i in range(k+1, n):
                    c[k] += mu[i][k]*v[i]
                c[k] *= -1
                v[k] = round(c[k])
                w[k] = 1
        else:
            k += 1
            if k == n:
                return s
            if k >= last_nonzero:
                last_nonzero = k
                v[k] += 1
            else:
                if v[k] > c[k]:
                    v[k] -= w[k]
                else:
                    v[k] += w[k]
                w[k] += 1

def find_length(v):
    result = 0
    for i in range(len(v)):
        result += v[i]**2
    
    return sqrt(result)

b = [[44.1, 50.99] ,[220.0, 44.0]]

lll = LLL(b)
r = sqrt(inner_product(b[0],b[0]))


l = schnorr_euchner(lll, r)
print(find_length(l))

File: bruh.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>
using namespace std;
using Vector = vector<double>;
using Basis = vector<Vector>;

double inner_product(const Vector u, const Vector v){

    double counter = 0;
    for (int i = 0; i<u.size(); ++i){
        counter += u[i]*v[i];
    } 
    return counter;
    }

Vector scalar_multiply(const Vector u, const double scalar) {
    
    Vector result;
    for (int i = 0; i < u.size(); ++i) {
        result.push_back(u[i] * scalar);
    }
    return result;
}

void print_vector(Vector v) {
    for (int i = 0; i < v.size(); ++i) {
        cout << v[i] << " ";
    }
    cout << "\n";
}

void print_matrix(Basis mat){
    for (const auto& row : mat) {
        for (auto elem : row) {
            cout << elem << ' ';
        }
        cout << '\n';
    }
}

Vector vector_subtraction(const Vector u, const Vector v) {
    Vector result;
    for (int i = 0; i < u.size(); ++i) {
        result.push_back(u[i]-v[i]);
    }
    return result;
}

Vector vector_add(const Vector u, const Vector v) {
    Vector result;
    for (int i = 0; i < u.size(); ++i) {
        result.push_back(u[i]+v[i]);
    }
    return result;
}

Basis gram_schmidt(const Basis A){
    Basis B;
    B.push_back(A[0]);

    for (int i = 1; i < A.size(); ++i){
        int counter = 0;
        Basis hats;
        while (counter < i) {
            double hat_top = inner_product(A[i], B[counter]);
            double hat_bottom = inner_product(B[counter], B[counter]);
            if (hat_bottom != 0){
            Vector hat_final = scalar_multiply(B[counter], hat_top/hat_bottom);
            hats.push_back(hat_final);
            }
            counter++;
        } 
        Vector result = A[i];
        for (int z = 0; z < hats.size(); ++z) {
            result = vector_subtraction(result, hats[z]);
        }
        B.push_back(result);
    }
    return B;
}

Basis LLL(Basis As, double S=0.75) {
    Basis A = As; // Buraya sonra kesin bak
    Basis B = gram_schmidt(A);
    int n = B.size();
    Basis mu(n, Vector(n,0)); // Bu ne aw
    for (int i = 0; i < n; ++i){
        for (int j = 0; j <= i; ++j) { // Buraya sonra bi bak
            mu[i][j] = inner_product(A[i], B[j])/inner_product(B[j], B[j]);
        }
    }

    int k = 1;
    while (k < n) {
        for (int j = k-1; j > -1; --j) {
            if (abs(mu[k][j]) > 0.5) {
                Vector tempp = scalar_multiply(A[j], round(mu[k][j]));
                A[k] = vector_subtraction(A[k], tempp);
                B = gram_schmidt(A);
                for (int i = 0; i < k+1; ++i) {
                    for (int p = 0; p <= i; ++p) {
                        mu[i][p] = inner_product(A[i], B[p])/inner_product(B[p],B[p]);
                    }
                
                }

            }
        }
        if (inner_product(B[k], B[k]) > (S-(pow((mu[k][k-1]),2))) * inner_product(B[k-1], B[k-1])){
            k++;
        }
        else {
            swap(A[k], A[k-1]);
            B = gram_schmidt(A);
            for (int i = 0; i < k+1; ++i) {
                for (int g = 0; g <= i; ++g) {
                    mu[i][g] = inner_product(A[i], B[g])/inner_product(B[g], B[g]);
                }
            }
            k = max(k-1, 1);
        }
    }
    return A;
}

int main(){
    Vector u = {1,2,3};
    Basis v = {{1, 1}, {1, 0}};
    Basis g = {{1, 2, 2}, {2, -1, 1}, {3, 0, 1}};
    Basis c = {{3, 1, 1}, {-1, 2, 4}, {2, 0, 3}};
    Basis deneme = {{201, 37}, {1648,297}};
    print_matrix(LLL(deneme));
    print_matrix(LLL(v));
    print_matrix(LLL(g));
    print_matrix(LLL(c));
    return 0;
}

